synapse_spamcheck_badlist/BadListFilter.py

```
@@ -0,0 +1,212 @@
+        if event["type"] != "m.room.message":
+            # We only filter messages.
+            return False
+
+        # If it's a text content, look for links.
+        if event["content"]["msgtype"] == "m.text":
+            if not await self.can_we_check_links():
+                return False
+
+            # Check for links in text, both unformatted and formatted.
+            #
+            # We always lower-case the text, as the IWF database is lowercase.
+            for text in [event.get("content", {}).get("body", ""), event.get("content", {}).get("formatted_body", "")]:
+                for match in self._linkifier.match(text.lower()) or []:
+                    link = re.sub(self._scheme_re, "", match.url)
+                    is_bad_link = await self._api.run_db_interaction("Check link against evil db", _db_is_bad_link, self._links_table, link)
+                    if is_bad_link:
+                        logger.info("Rejected bad link")
+                        return True
+
+        # If it's a file, download content, extract hash.
+        elif event["content"]["msgtype"] == "m.file":
+            if not await self.can_we_check_md5():
+                return False
+
+            match = self._mxc_re.match(event["content"]["url"])
+            server_name = None
+            media_id = None
+            if match != None:
+                server_name = match.group('server_name')
+                media_id = match.group('media_id')
+                response = None
+                try:
+                    url = "%s/_matrix/media/r0/download/%s/%s" % (
+                            self._base_url,
+                            urlquote(server_name),
+                            urlquote(media_id)
+                    )
+                    response = await self._api.http_client.request("GET", url)
+                    # FIXME: What do we do if we were rate-limited?
+                except Exception as e:
```

This error checking does not seem to work, I think you want to `continue` after logging.

---

I think we want to bailout and `return False`, no?

---

I don't know what behavior is best, it depends if you want to fail closed or opened (and potentially if you want retries). You might also want to check other links.

---

Other links? There can be more than one file in a `m.file` message?

---

Oops, sorry I got this confused with checking the links in a standard message!